c10/core/DispatchKey.h

```diff
@@ -328,6 +330,10 @@ enum class DispatchKey : uint8_t {
   // If you add new alias keys after Autograd, please also update it here.
   EndOfAliasKeys = CompositeExplicitAutograd, //
 
+  // The CsrTranspose dispatch key is set for any SparseCsr tensor that
+  // is a sparse CSC tensor
+  CsrTranspose,
```

Comment:
Hm. @ezyang - should a new layout use a new dispatch key?

Reply:
This is exactly the place where I'd like to get early feedback.

Initially, I tried adding `CsrTranspose` after `ZeroTensor` but that hit immediately the 64 dispatch keys limitation, so moved it under the alias keys section.

Re the question: I think this is related to
https://github.com/pytorch/pytorch/blob/48b7e60835f50c3735d54bf886c00f1340c82e30/c10/core/TensorImpl.h#L1001-L1014
that is, the layout is derived from the dispatch keys and the plan was to introduce the CSC layout without introducing new dispatch keys such as SparseCSCCPU/CUDA keys.

Reply:
Once I'm able to land https://github.com/pytorch/pytorch/pull/72827 we'll have more dispatch key space, so you'll be able to add a `CsrTranspose` key if we need it.

Reply:
@pearu - can we do this without introducing a new dispatch key for now? We can add a "is_transposed" bool to SparseCsrTensorImpl to figure out whether a given instance is CSR or CSC. We can then add a new dispatch key after @bdhirsh finished the work.

Reply:
@cpuhrsch when we add "is_transposed" bool to SparseCsrTensorImpl and we still want to support CSC layout, we'll have a problem with defining "is_sparse_csr/csc" methods in TensorImpl: these methods are assumed to be non-virtual (so when being inside these methods, there is no glue if a particular tensors instance is SparseCsrTensorImpl or XYZTensorImpl instance) and these are needed in the (non-virtual) "layout()" method of TensorImpl. IIUC the torch tensor design, defining a new layout without defining a dispatch key that would relate to the layout in TensorImpl (non-virtual) methods, is simply not possible.

On the other hand, what is the problem that we could not define the new dispatch key now? Although it is not a "runtime" dispatch key, the current implementation already allows differentiating the CSR and CSC tensors in TensorImpl non-virtual methods. When @bdhirsh work lands, we would only need to move the CsrTranspose dispatch key to the section of "runtime" dispatch keys to make CsrTranspose a "runtime" dispatch key.

Reply:
side note: the above PR is hard constraint for adding a new dispatch key (you won't be able to use an "alias dispatch key" and get anything to work). Alias dispatch keys are basically just logical groupings of runtime dispatch keys (e.g. the `Autograd` alias key maps to `AutogradCPU/CUDA/...` runtime keys), and they don't get assigned any runtime bits. You can't do things like `keyset.has(DispatchKey::CsrTranspose)` if it's an alias key.

Reply:
@pearu - ok then it seems more reasonable to first continue onto other work items and wait until @bdhirsh's work has landed.

Reply:
you can always kill an existing key to prototype though :)

I'll let you know when the PR to free up space lands - I'm trying to push on it this week, and it's sitting here: https://github.com/pytorch/pytorch/pull/72827

Reply:
Now using #72827 . Resolving as obsolete.